bindings/python/src/ucp_py.pyx

```diff
@@ -0,0 +1,488 @@
+
+    _instances = WeakValueDictionary()
+
+    def __init__(self, ucp_msg = None):
+        self.done_state = False
+        self.result_state = None
+        #self.start_time = time.time()
+        #self.end_time = None
+        self._instances[id(self)] = self
+        if None != ucp_msg:
+            self.ucp_msg = ucp_msg
+        super(CommFuture, self).__init__()
+
+    def done(self):
+        if False == self.done_state and hasattr(self, 'ucp_msg'):
+            if 1 == self.ucp_msg.query():
+                self.done_state = True
+                #self.end_time = time.time()
+                #lat = self.end_time - self.start_time
+                #print("future time {}".format(lat * 1000000))
+                self.result_state = self.ucp_msg
+                self.set_result(self.ucp_msg)
+        return self.done_state
+
+    def result(self):
+        while False == self.done_state:
+            self.done()
+        return self.result_state
+
+    def __del__(self):
+        pass #self.ucp_msg.free_mem()
+
+    def __await__(self):
+        if True == self.done_state:
+            return self.result_state
+        else:
+            while False == self.done_state:
+                if True == self.done():
+                    return self.result_state
+                else:
+                    yield
+
+class ListenerFuture(concurrent.futures.Future):
+    """A class to keep listener alive and invoke callbacks on incoming
+    connections
+    """
+
+    _instances = WeakValueDictionary()
+
+    def __init__(self, cb):
+        self.done_state = False
+        self.result_state = None
+        self.cb = cb
+        self._instances[id(self)] = self
+        super(ListenerFuture, self).__init__()
+
+    def done(self):
+        if False == self.done_state:
+            ucp_py_worker_progress()
+        return self.done_state
+
+    def result(self):
+        while False == self.done_state:
+            self.done()
+        return self.result_state
+
+    def __del__(self):
+        pass
+
+    def __await__(self):
+        if True == self.done_state:
+            return self.result_state
+        else:
+            while False == self.done_state:
+                if True == self.done():
+                    return self.result_state
+                else:
+                    yield
+
+cdef class ucp_py_ep:
+    """A class that represents an endpoint connected to a peer
+    """
+
+    cdef void* ucp_ep
+    cdef int ptr_set
+
+    def __cinit__(self):
+        return
+
+    def connect(self, ip, port):
+        self.ucp_ep = ucp_py_get_ep(ip, port)
+        return
+
+    def recv_future(self):
+        """Blind receive operation"""
+
+        recv_msg = ucp_msg(None)
+        recv_msg.ucp_ep = self.ucp_ep
+        recv_future = CommFuture(recv_msg)
+        ucp_py_ep_post_probe()
+        return recv_future
+
+    def recv(self, ucp_msg msg, len):
+        """Receive operation of length `len`
```

Comment:
Please stick with Python's docstring conventional structure throughout the code.